### ISSUE TITLE
Develop and implement guidelines with regards to handling startup problems

### ISSUE THREAD

**Issue by dylandreimerink** [cilium/cilium#29169]
Within the agent (and perhaps the operator as well) there are some major inconsistencies regarding how we handle certain situations during startup. The situations I have been able to identify are:
* If config X is set to A, then config Y must be C or D
* If config X is set to A then config Y may not be C
* If feature X is enabled, then during init we probe for Z, and the probe is negative
* If feature X is enabled, then during init we change external settings (sysctl, fs, iptables) and this errors

There might be others, please add them if so.

There are a number of ways to handle these (combinations possible):
* Return a error, shutting down the agent, causing a crash backoff loop
* Changing configuration (Change Y to B if X is set to A)
* Fallback by implementing an inferior version (not always possible)
* "Fallback" by disabling a feature
* Attempt to reconcile after init
* Make a log message at some level
* Mark the component as degraded in the health reporter

We currently lack a formal policy or guideline that states how we want to handle which situations. We should discuss this attempt to formulate some sort of guidance. Then attempt to implement it as consistently as possible and refine where needed.

**Comment by bimmlerd** [comment]
Some thoughts:

1. Cilium currently lacks a differentiation between user configuration (intent) and runtime configuration (reality). To me, the intent of the user should be immutable, and the runtime configuration would be a derivation of it, with the additional input of feature probing/system state. 
2. If cilium cannot satisfy the intent of the user (i.e. would override the configuration due to the realities of where it's running), I believe it should crash, loudly and with a helpful error message. For features which depend on new kernel version, where we implement fallbacks, I'd do so only with a `--allow-x-fallback`, i.e. a feature gate, so that the choice is a conscious one.

Now, those are idealistic principles which clash badly with reality. The configuration of the agent is (very) complicated and doesn't offer a good user experience. As an example, it's not like we can say `--encryption={wireguard,ipsec}` - instead, we have to enable either, and the flag names aren't exactly intuitive.

**Comment by tommyp1ckles** [comment]
> The situations I have been able to identify are:

One issue with this is that it's hard to know what the "rules" are without reading the code.  Whatever the decision tree for this ends up being it'd be nice if we could have some way to encode this in a single place to make it easy to understand what the rule is, how one might change this (i.e. disable probing?).  Rather than having the rules be various if/else logic or how we decide to handle errors in some situations.

> If cilium cannot satisfy the intent of the user (i.e. would override the configuration due to the realities of where it's running), I believe it should crash, loudly and with a helpful error message. For features which depend on new kernel version, where we implement fallbacks, I'd do so only with a --allow-x-fallback, i.e. a feature gate, so that the choice is a conscious one.

I tend to agree with this,  we should make it easy to understand what the problem & resolution are rather than making it easy to run incompatible configuration.

As mentioned in the community meeting however, the exception to this seems to be providing upgrade/downgrade/configuration paths. However those should likely be opt-in (inevitably those cases are probably not going to be as well tested as the general path approach).

**Comment by dylandreimerink** [comment]
> The configuration of the agent is (very) complicated and doesn't offer a good user experience

> we should make it easy to understand what the problem & resolution are rather than making it easy to run incompatible configuration.

> One issue with this is that it's hard to know what the "rules" are without reading the code.

I agree with all of this. I believe this also matches what @derailed said in the community meeting about how it should not be possible to pass invalid configuration to the agent.

I think its clear that we need to improve our current way of doing config at multiple levels, instead of trying to do what this issue originally stated. I think I will take the CFP route for that.

I wanted to do a quick brain dump first.

* Cilium has quite a few ways of configuring it (ENV variables, ConfigMap, flags, Per node config) so we will likely need to enforce correctness at a few levels
* The most common workflow is to use Helm or Cilium CLI (Helm underneath) to manage configuration. Helm has the ability to do config validation https://github.com/cilium/cilium/blob/main/install/kubernetes/cilium/templates/validate.yaml
* Having a "ruleset" as basis, from which we can generate validation logic in Go, validation logic in Helm templates, and hopefully something human readable would be nice
* Perhaps include a TUI config util in Cilium CLI based on the rule set of to be installed cilium version?
* Failable config options should include user intent. `kernel-dependent-feature: EnableIfAvailable | MustEnable | Disabled`
* Mutually exclusive options should be enums `tunnel-mode: disabled | vxlan | geneve | ipip` where ever possible

Then in the agent, we would crash/shutdown whenever conflicting config is detected (should only happen in validation is bypassed or broken). For probe failure or fallback we look at the communicated intent, and only crash if they communicated they wanted that rater than disabling or running in a fallback mode.
